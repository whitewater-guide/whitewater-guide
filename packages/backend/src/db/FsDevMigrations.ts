// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-ignore
import { FsMigrations } from 'knex/lib/migrate/sources/fs-migrations';

// When restoring production dump on development, migration names will have '.js' extension
// Therefore, development migrations with '.ts' will conflict them
// this is the workaround
// https://github.com/knex/knex/blob/master/lib/migrate/sources/fs-migrations.js
export class FsDevMigrations extends FsMigrations {
  // eslint-disable-next-line @typescript-eslint/no-useless-constructor
  constructor(
    migrationDirectories: string | string[],
    sortDirsSeparately?: boolean,
    loadExtensions?: string[],
  ) {
    super(migrationDirectories, sortDirsSeparately, loadExtensions);
  }

  getMigrationName(migration: any): string {
    return migration.file.replace('.ts', '.js');
  }
}
