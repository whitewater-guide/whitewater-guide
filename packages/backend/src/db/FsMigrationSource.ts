import fs from 'node:fs/promises';
import path from 'node:path';

import type { Knex } from 'knex';

// https://knexjs.org/guide/migrations.html#custom-migration-sources
export default class FsMigrationSource implements Knex.MigrationSource<string> {
  // eslint-disable-next-line node/no-process-env
  private readonly ext = process.env.NODE_ENV === 'production' ? 'js' : 'ts';
  private readonly directory: string;

  constructor(directory: string) {
    this.directory = directory;
  }
  // Must return a Promise containing a list of migrations.
  // Migrations can be whatever you want,
  // they will be passed as arguments to getMigrationName
  // and getMigration
  //
  // In our case migrations are filenames without extension
  async getMigrations(): Promise<string[]> {
    // In this example we are just returning migration names
    const dirContent = await fs.readdir(this.directory);
    // assume that migrations have extensions (== no directories with '.' symbol)
    return dirContent
      .filter((f) => f.includes('.'))
      .map((f) => f.split('.')[0])
      .sort();
  }

  // When restoring production dump on development, migration names will have '.js' extension
  getMigrationName(migration: string): string {
    return migration + '.js';
  }

  async getMigration(migration: string): Promise<Knex.Migration> {
    const moduleName = path.resolve(
      process.cwd(),
      this.directory,
      `${migration}.${this.ext}`,
    );
    return import(moduleName);
  }
}
