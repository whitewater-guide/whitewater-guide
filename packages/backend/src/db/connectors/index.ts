export * from './createConnectors';
export * from './OffsetConnector';
export * from './types';
