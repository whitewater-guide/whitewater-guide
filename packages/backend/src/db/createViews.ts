import { readFileSync } from 'fs';
import Knex from 'knex';
import path from 'path';

export async function createView(db: Knex, view: string, version = 1) {
  for (let i = version; i >= 1; i--) {
    const file = path.resolve(
      __dirname,
      '../migrations',
      `${i.toString(10).padStart(3, '0')}/${view}_view.sql`,
    );
    let sql: string;
    try {
      sql = readFileSync(file).toString();
    } catch (e) {
      continue;
    }
    await db.schema.raw(sql);
    return;
  }
  throw new Error(
    `No sql files for view '${view}' of version ${version} or lower found`,
  );
}

// Views will be created in reverse order
export async function createViews(
  db: Knex,
  version: number,
  ...views: string[]
) {
  const inverted = [...views].reverse();
  for (const view of inverted) {
    await createView(db, view, version);
  }
}
