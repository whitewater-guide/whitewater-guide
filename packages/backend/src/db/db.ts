/* eslint-disable node/no-process-env */
import Knex from 'knex';

import knexConfig from './knexfile';

const env = process.env.NODE_ENV || 'development';
const config: Knex.Config = knexConfig[env];

export const knex = Knex(config);

let testTransaction: Knex.Transaction | null = null;

export function db(getKnexInstance = false): Knex<any, any> {
  if (env === 'test' && !getKnexInstance) {
    if (testTransaction) {
      return testTransaction;
    }
    throw new Error('No saved test transaction found');
  }
  return knex;
}

// See here https://github.com/tgriesser/knex/issues/2076
// And here https://gist.github.com/odigity/7f37077de74964051d45c4ca80ec3250
export const holdTransaction = () => {
  if (env !== 'test') {
    throw new Error('holdTransaction is only available in test mode');
  }
  const tmp: any = {};
  const p = new Promise((resolve) => {
    tmp.resolve = resolve;
  });
  knex
    .transaction((tx) => {
      testTransaction = tx;
      tmp.resolve();
    })
    .catch(() => {
      // ignore
    });
  return p;
};

export const rollbackTransaction = async () => {
  if (env !== 'test') {
    throw new Error('rollbackTransaction is only available in test mode');
  }
  if (testTransaction) {
    await testTransaction.rollback();
    testTransaction = null;
  }
};
