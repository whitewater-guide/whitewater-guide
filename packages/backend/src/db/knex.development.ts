import knex from 'knex';

const development: knex.Config = {
  client: 'pg',
  connection: {
    host: process.env.POSTGRES_HOST!,
    port: 5432,
    database: process.env.POSTGRES_DB!,
    user: 'postgres',
    password: process.env.POSTGRES_PASSWORD!,
    keepAlive: true,
  },
  migrations: {
    loadExtensions: ['.js'],
    tableName: 'migrations',
    directory: './src/migrations',
  },
  seeds: {
    loadExtensions: ['.js'],
    directory: './src/seeds/development',
  },
  debug: process.env.DATABASE_DEBUG === 'true',
} as any;

module.exports = development;
