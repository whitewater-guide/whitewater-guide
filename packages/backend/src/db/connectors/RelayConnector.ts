import type { RelayConnection } from '@whitewater-guide/schema';
import type { Knex } from 'knex';
import { clamp } from 'lodash';

import type { Cursor } from '../../apollo/index';
import { db } from '../../db/index';
import { BaseConnector } from './BaseConnector';
import type { ManyBuilderOptions } from './types';

export abstract class RelayConnector<
  TGraphql,
  TSql extends { id: string },
> extends BaseConnector<TGraphql, TSql> {
  protected _defaultPageSize = 25;

  protected _maxPageSize = 100;

  protected abstract getAfterClause(cursor: Cursor): string;

  public getManyQuery(
    options: ManyBuilderOptions<TSql> = {},
  ): Knex.QueryBuilder {
    let query = this.buildGenericQuery().select(db().raw(`count(*) OVER()`));

    // Simple filtering
    if (options.where) {
      query = query.where(options.where);
    }

    // Offset pagination
    if (options.page?.limit) {
      query = query.limit(options.page.limit);
      if (options.page.offset) {
        query = query.offset(options.page.offset);
      }
    }

    // Limit
    const limit = clamp(
      options.page?.limit || this._defaultPageSize,
      1,
      this._maxPageSize,
    );
    query = query.limit(limit);
    // Relay-style pagination
    const after = options.page?.after;
    if (after) {
      query = query.whereRaw(this.getAfterClause(after));
    }

    // Sort
    const orderBy = options.orderBy || this._orderBy;
    if (orderBy) {
      orderBy.forEach(({ column, direction }) => {
        query = query.orderBy(column, direction);
      });
    }

    return query;
  }

  public itemsToConnection<TItem = any>(
    items: any[] | undefined,
    total: number,
    orderBy?: string,
  ): RelayConnection<TItem> {
    const edges =
      items?.map((node) => {
        let value = orderBy ? node[orderBy] : undefined;
        if (value instanceof Date) {
          value = value.getTime().toString(10);
        }
        return {
          cursor: {
            ordId: node.ord_id,
            value,
          },
          node,
        };
      }) || [];
    const hasMore = edges.length < total;
    if (!edges.length) {
      return { edges, pageInfo: { hasMore, endCursor: null } };
    }
    const endEdge = edges[edges.length - 1];
    return { edges, pageInfo: { hasMore, endCursor: endEdge.cursor } };
  }
}
