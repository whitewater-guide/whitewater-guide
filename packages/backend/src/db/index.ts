import db from './db';

export default db;
export * from './updatedAtTrigger';
export * from './runSqlFile';
export * from './createViews';
export * from './dropViews';
export * from './rawUpsert';
export * from './types';
export * from './db';
export * from './queryBuilders';
