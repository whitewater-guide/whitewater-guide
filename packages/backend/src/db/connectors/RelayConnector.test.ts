import type { Cursor } from '../../apollo/index';
import { Context } from '../../apollo/index';
import { holdTransaction, rollbackTransaction } from '../../db/index';
import { RelayConnector } from './RelayConnector';
import type { FieldsMap } from './types';

beforeAll(holdTransaction);
afterAll(rollbackTransaction);

class TestConnector extends RelayConnector<any, any> {
  constructor() {
    super(new Context());
    this._tableName = 'test_table';
    this._graphqlTypeName = 'TestType';
    this._language = undefined;
  }

  protected getAfterClause(cursor: Cursor): string {
    return `(birthday, ord_id) < (${cursor.value}, ${cursor.ordId})`;
  }

  public fieldsByType(fields: string[]): this {
    this._fieldsByType = new Map([[this._graphqlTypeName, new Set(fields)]]);
    return this;
  }

  public fieldsMap(map: FieldsMap<any, any>): this {
    this._fieldsMap = map;
    return this;
  }

  public sqlFields(fields: string[]): this {
    this._sqlFields = Array.from(new Set([...this._sqlFields, ...fields]));
    return this;
  }

  public language(lng: string): this {
    this._language = lng;
    return this;
  }
}

it('should build correct query', () => {
  const query = new TestConnector()
    .fieldsByType(['birthday', 'name'])
    .sqlFields(['ord_id'])
    .getManyQuery({
      where: { name: 'foo' },
      page: { limit: 3, after: { value: '2020-01-01', ordId: 4 } },
      orderBy: [
        {
          column: 'birthday',
          direction: 'desc',
        },
        {
          column: 'ord_id',
          direction: 'desc',
        },
      ],
    });
  expect(query).toMatchSnapshot();
});

it('should produce relay-style connection', () => {
  const connector = new TestConnector();
  const actual = connector.itemsToConnection(
    [
      {
        name: 'foo',
        created_at: '1',
        ord_id: 1,
      },
      {
        name: 'bar',
        created_at: '2',
        ord_id: 2,
      },
    ],
    10,
    'created_at',
  );
  expect(actual).toEqual({
    edges: [
      {
        cursor: {
          ordId: 1,
          value: '1',
        },
        node: {
          created_at: '1',
          name: 'foo',
          ord_id: 1,
        },
      },
      {
        cursor: {
          ordId: 2,
          value: '2',
        },
        node: {
          created_at: '2',
          name: 'bar',
          ord_id: 2,
        },
      },
    ],
    pageInfo: {
      endCursor: {
        ordId: 2,
        value: '2',
      },
      hasMore: true,
    },
  });
});

it('should produce relay-style connection for date values', () => {
  const connector = new TestConnector();
  const actual = connector.itemsToConnection(
    [
      {
        name: 'foo',
        created_at: new Date(2010, 1, 1),
        ord_id: 1,
      },
      {
        name: 'bar',
        created_at: new Date(2012, 1, 1),
        ord_id: 2,
      },
    ],
    10,
    'created_at',
  );
  expect(actual).toEqual({
    edges: [
      {
        cursor: {
          ordId: 1,
          value: '1264982400000',
        },
        node: {
          created_at: new Date(2010, 1, 1),
          name: 'foo',
          ord_id: 1,
        },
      },
      {
        cursor: {
          ordId: 2,
          value: '1328054400000',
        },
        node: {
          created_at: new Date(2012, 1, 1),
          name: 'bar',
          ord_id: 2,
        },
      },
    ],
    pageInfo: {
      endCursor: {
        ordId: 2,
        value: '1328054400000',
      },
      hasMore: true,
    },
  });
});
