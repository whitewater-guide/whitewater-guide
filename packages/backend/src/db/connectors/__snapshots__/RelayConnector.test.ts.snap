// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should build correct query 1`] = `
select
    id,
    ord_id,
    birthday,
    name,
    count(*) OVER ()
from
    test_table
where
    name = 'foo'
    and (birthday,
        ord_id) < (2020 - 01 - 01,
        4)
order by
    birthday desc,
    ord_id desc
limit 3

`;
